datalad/core/distributed/clone.py

```diff
@@ -636,3 +691,101 @@ def _get_installationpath_from_url(url):
+    # standard property dict composition
+    props = dict(
+        source=spec,
+        version=None,
+    )
+    # common starting point is a RI instance, support for accepting an RI
+    # instance is kept for backward-compatibility reasons
+    source_ri = RI(spec) if not isinstance(spec, RI) else spec
+
+    # scenario switch, each case must set 'giturl' at the very minimum
+    if isinstance(source_ri, DataLadRI):
+        # we have got our DataLadRI as the source, so expand it
+        props['type'] = 'dataladri'
+        props['giturl'] = source_ri.as_git_url()
+    elif isinstance(source_ri, URL) and source_ri.scheme.startswith('ria+'):
+        # parse a RIA URI
+        # check for store related configuration for SSH access
+        if source_ri.scheme == 'ria+ssh':
+            # we support the scenario were hostname is just a label for a config
+            # not an actual hostname
+            hostname = cfg.get(
+                'annex.ria-remote.{}.ssh-host'.format(source_ri.hostname),
+                source_ri.hostname)
+            basepath = cfg.get(
+                'annex.ria-remote.{}.base-path'.format(source_ri.hostname),
+                '')
+        else:
+            hostname = source_ri.hostname
+            basepath = ''
+
+        dsid, version = source_ri.fragment.split('@', maxsplit=1) \
+            if '@' in source_ri.fragment else (source_ri.fragment, None)
+        uuid_regex = r'[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}'
+        if not re.match(uuid_regex, dsid):
+            raise ValueError('RIA URI does not contain a valid dataset ID: {}'.format(spec))
+        # now we cancel the fragment in the original URL, but keep everthing else
+        # in order to be able to support the various combinations of ports, paths,
+        # and everything else
+        source_ri.fragment = ''
+        # strip the custom protocol and go with standard one
+        source_ri.scheme = source_ri.scheme[4:]
+        # take any existing path, and add trace to dataset within the store
+        source_ri.path = '{urlpath}{urldelim}{basepath}{basedelim}{trace}'.format(
+            urlpath=source_ri.path if source_ri.path else '',
+            urldelim='' if not source_ri.path or source_ri.path.endswith('/') else '/',
+            basepath=basepath,
+            basedelim='' if not basepath or basepath.endswith('/') else '/',
+            trace='{}/{}'.format(dsid[:3], dsid[3:]),
+        )
+        props.update(
+            type='ria',
+            giturl=str(source_ri),
+            version=version,
+            default_destpath=dsid,
+        )
+    else:
+        # let's assume that anything else is a URI that Git can handle
+        # TODO are there any advantages of going through a URL-reconversion
+        # by the RI instance, instead of using the original input right away?
+        # it feels like we could only accumulate bugs. Keeping how it was,
+        # for now...
```

Comment:
I agree that it'd be better to use the original value.  This is particularly important because git supports flexible sources via `gitremote-helper`.

By the way, in the context of the call's discussion today on whether "ria" (or whatever name) should be in the scheme part or in the fragment, in my view `gitremote-helper` provides two related points in favor of putting it in the scheme: (1) to be consistent with what git does (`<transport>://<address>` or `<transport>::<address>`) and (2) it keeps open the possibility of eventually implementing remote helpers for `ria+` URLs so that git commands can work directly with them.

Reply:
Thank you @kyleam for pointing to `gitremote-helper` -- I should have remembered about it since I must have looked at it in the past.
My non-executive summary:
- I am Ok with going with this PR as is ATM

but in the long run (although `datalad::ria::` instead of `ria+` could be a short one) I think
- we should RF to provide `datalad` git remote helper and advise to use `datalad::url` or at least to adhere to such schema ourselves to start. **pros**: even for local paths it would become sensible `datalad::some/possibly/relative/path` (what would it be now in this PR?  haven't seen an example, could it would be smth like `ria+some/possibly/relative/path` if to stay consistent). That would make it clear to the user who sees `datalad::` prefix that it relates to datalad, similarly to how `hg::URL` gives immediate information to me that remote URL I am giving to `git clone` is `hg` repository. (please no grandma examples, use prototypical datalad users)
- layout could be specified as explicit option in the fragment, e.g. `#layout=uuid2` (and describe in docs what uuid2 is and that ria uses such layout scheme (hardcoded)) e.g. `datalad::path#id=UUID;layout=uuid2` or `datalad::path#layout=dataset-tarballs` (for whenever we implement export of the datasets with tarball for `.git` itself), but   
- I think that we might want to altogether avoid hard coding assumptions of the layout on remote end into URL, be it a `ria` (abbreviation which has nothing to do with layout per se) or anything else; neither in prefix of transport nor in URL fragment like I was proposing.  I think, we might want it to be "queried" by datalad from that remote (e.g. some config option within `.datalad/config` on the remote end) similarly how to git-annex queries UUID etc.  It would though require providing support for underlying transport mechanisms to requests such a file (or just entire super/top dataset to some temporary location if not present), but **pros**: it would open possibilities for migrating layout/etc on remote end without requiring adjustment to the URLs or code, and avoid possible guesswork then otherwise needed.  But ATM we even do guesswork for either there is `/.git` or not, so I guess it would be a long shot.
- may alternatively in the short run we could/should "chain" hardcoded `ria` as `datalad::ria::URL` but I am not found of it as much as of `ria+` because then no alternative tool (e.g. browser rendering top JS page of the URL, such as http://store.datalad.org) to datalad would be given a chance.

Reply:
> ria (abbreviation which has nothing to do with layout per se) 

Yes and no. You can have that layout without a RIA special remote and current implementation here does that. But: The other way around: RIA would produce such a layout. So, if you know your store was created by it, you wouldn't need know about the layout but just that it's a RIA(-created) store.
Another aspect of it (not currently implemented here): That layout RIA special remote is created is versioned and stored in that store's base-path. So, if we know it's a RIA store and the special remote makes it into core, we can query the special remote for the correct layout without changing that URL. RIA remote's own implementation does exactly that. To do it here would require to import it, of course.


>  I think, we might want it to be "queried" by datalad from that remote (e.g. some config option within .datalad/config on the remote end) similarly how to git-annex queries UUID etc.

I actually wanted to make a very similar point as an argument for using the scheme rather than fragment: Currently we can have a label for a store, that just determines the base of the dataset tree as an actual URL. But: If it's not a regular URL we can enhance that to adress a distrubuted store. That is, several of such stores the user would know about as one thing and let datalad query an index to figure out which store to use for a given dataset id.